src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -33,5 +155,874 @@ pub(super) fn persist_sink<S>(
 where
     S: Scope<Timestamp = Timestamp>,
 {
-    todo!()
+    let mut scope = ok_collection.scope();
+    let desired = OkErr::new(ok_collection.inner, err_collection.inner);
+
+    // Read back the persist shard.
+    let (mut persist, persist_token) = persist_source(
+        &mut scope,
+        sink_id,
+        target.clone(),
+        compute_state,
+        start_signal,
+    );
+
+    // Report sink frontier updates to the `ComputeState`.
+    let sink_frontier = Rc::new(RefCell::new(Antichain::from_elem(Timestamp::MIN)));
+    let collection = compute_state.expect_collection_mut(sink_id);
+    collection.sink_write_frontier = Some(Rc::clone(&sink_frontier));
+
+    persist.ok = persist.ok.inspect_container(move |event| {
+        if let Err(frontier) = event {
+            let mut borrow = sink_frontier.borrow_mut();
+            borrow.clear();
+            borrow.extend(frontier.iter().copied());
+        }
+    });
+
+    // Determine the active worker for single-worker operators.
+    let active_worker_id = usize::cast_from(sink_id.hashed()) % scope.peers();
+
+    let operator_name = |name| format!("persist_sink({sink_id})::{name}");
+    let persist_api = |purpose| PersistApi {
+        persist_clients: Arc::clone(&compute_state.persist_clients),
+        collection: target.clone(),
+        shard_name: sink_id.to_string(),
+        purpose,
+    };
+
+    let name = operator_name("mint");
+    let (desired, descs, mint_token) = mint::render(
+        name.clone(),
+        persist_api(name),
+        as_of,
+        active_worker_id,
+        &desired,
+    );
+
+    let name = operator_name("write");
+    let (batches, write_token) =
+        write::render(name.clone(), persist_api(name), &desired, &persist, &descs);
+
+    let name = operator_name("append");
+    let append_token = append::render(
+        name.clone(),
+        persist_api(name),
+        active_worker_id,
+        &descs,
+        &batches,
+    );
+
+    Rc::new((persist_token, mint_token, write_token, append_token))
+}
+
+/// Generic wrapper around ok/err pairs (e.g. streams, frontiers), to simplify code dealing with
+/// such pairs.
+struct OkErr<O, E> {
```

Comment:
I'm not sure I like the `OrErr` struct. I see that it bundles ok and error variants in several places, but does it add more than avoiding additional variables? For me as a reader, it requires extra context to understand the implementation below.

Reply:
Not really, it only cleans up the usage sites and removes some duplication. Most significantly it encapsulates getting a global frontier from two tracked oks/errs frontiers, which is something we have to do in a couple of places.

Reply:
We did this initially and changed to the current demultiplexed approach in https://github.com/MaterializeInc/materialize/pull/26779. As the PR states, this is to save 8 bytes of memory for each record, which is significant.